### PR TITLE
Use spans instead of ptr and size in kernel memory

### PR DESCRIPTION
**Pull request by MCredstoner2004**
This makes the code more readable and allowed removing a bunch of one-liner functions

### CODE REVIEW

File: app/src/main/cpp/skyline/kernel/memory.cpp [hunk 1]
```diff
@@ -24,16 +24,14 @@ namespace skyline::kernel {
                 throw exception("32-bit address spaces are not supported");
 
             case memory::AddressSpaceType::AddressSpace36Bit: {
-                addressSpace.address = 0;
-                addressSpace.size = 1UL << 36;
-                base.size = 0x78000000 + 0x180000000 + 0x78000000 + 0x180000000;
+                addressSpace = span<u8> {static_cast<u8 *>(nullptr), 1UL << 36};
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/memory.cpp:27]: don't do this, best to do size seperate and add ptr to construct a span after

File: app/src/main/cpp/skyline/kernel/memory.cpp [hunk 2]
```diff
@@ -49,115 +47,103 @@ namespace skyline::kernel {
...
-        if (address < base.address || address + size > base.address + base.size)
-            throw exception("Mapping is outside of VMM base: 0x{:X} - 0x{:X}", address, address + size);
+    span<u8> MemoryManager::CreateMirror(span<u8> mapping) {
+        if (mapping.data() < base.data() || mapping.end().base() > base.end().base())
+            throw exception("Mapping is outside of VMM base: 0x{:X} - 0x{:X}", mapping.data(), mapping.end().base());
 
-        size_t offset{address - base.address};
-        if (!util::IsPageAligned(offset) || !util::IsPageAligned(size))
-            throw exception("Mapping is not aligned to a page: 0x{:X}-0x{:X} (0x{:X})", address, address + size, offset);
+        auto offset{static_cast<size_t>(mapping.data() - base.data())};
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/memory.cpp:138]: cast not needed

File: app/src/main/cpp/skyline/kernel/memory.cpp [hunk 3]
```diff
@@ -171,17 +157,16 @@ namespace skyline::kernel {
...
         size_t mirrorOffset{};
         for (const auto &region : regions) {
-            auto address{reinterpret_cast<u64>(region.data())};
-            if (address < base.address || address + region.size() > base.address + base.size)
-                throw exception("Mapping is outside of VMM base: 0x{:X} - 0x{:X}", address, address + region.size());
+            if (region.data() < base.data() || region.end().base() > base.end().base())
+                throw exception("Mapping is outside of VMM base: 0x{:X} - 0x{:X}", region.data(), region.end().base());
 
-            size_t offset{address - base.address};
+            size_t offset{static_cast<size_t>(region.data() - base.data())};
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/memory.cpp:163]: cast not needed

File: app/src/main/cpp/skyline/kernel/types/KPrivateMemory.cpp [hunk 4]
```diff
@@ -8,66 +8,63 @@
...
-        if (mprotect(nPtr, nSize, PROT_READ | PROT_WRITE | PROT_EXEC) < 0)
+        if (mprotect(map.data(), map.size(), PROT_READ | PROT_WRITE | PROT_EXEC) < 0)
             throw exception("An occurred while remapping private memory: {}", strerror(errno));
     }
 
-    void KPrivateMemory::UpdatePermission(u8 *pPtr, size_t pSize, memory::Permission pPermission) {
-        pPtr = std::clamp(pPtr, ptr, ptr + size);
-        pSize = std::min(pSize, static_cast<size_t>((ptr + size) - pPtr));
+    void KPrivateMemory::UpdatePermission(span<u8> map, memory::Permission pPermission) {
+        auto pPtr = std::clamp(map.data(), guest.data(), guest.end().base());
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/types/KPrivateMemory.cpp:66]: drop p and use {}, p is for parameters and it's no longer from a parameter henc enot needed

File: app/src/main/cpp/skyline/kernel/types/KProcess.cpp [hunk 5]
```diff
@@ -46,7 +46,7 @@ namespace skyline::kernel::type {
 
     void KProcess::InitializeHeapTls() {
         constexpr size_t DefaultHeapSize{0x200000};
-        heap = std::make_shared<KPrivateMemory>(state, reinterpret_cast<u8 *>(state.process->memory.heap.address), DefaultHeapSize, memory::Permission{true, true, false}, memory::states::Heap);
+        heap = std::make_shared<KPrivateMemory>(state, span<u8>(reinterpret_cast<u8 *>(state.process->memory.heap.data()), DefaultHeapSize), memory::Permission{true, true, false}, memory::states::Heap);
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/types/KProcess.cpp:49]: use {} init for all spans

File: app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp [hunk 6]
```diff
@@ -10,76 +10,76 @@
...
             throw exception("An error occurred while creating shared memory: {}", fd);
 
-        host.ptr = static_cast<u8 *>(mmap(nullptr, size, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_SHARED, fd, 0));
-        if (host.ptr == MAP_FAILED)
+        auto host_ptr = static_cast<u8 *>(mmap(nullptr, size, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_SHARED, fd, 0));
+        if (host_ptr == MAP_FAILED)
             throw exception("An occurred while mapping shared memory: {}", strerror(errno));
 
-        host.size = size;
+        host = span<u8>(host_ptr, size);
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp:22]: {}

File: app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp [hunk 7]
```diff
@@ -10,76 +10,76 @@
...
+        if (!state.process->memory.base.contains(map))
+            throw exception("KPrivateMemory allocation isn't inside guest address space: 0x{:X} - 0x{:X}", map.data(), map.end().base());
+        if (!util::IsPageAligned(map.data()) || !util::IsPageAligned(map.size()))
+            throw exception("KSharedMemory mapping isn't page-aligned: 0x{:X} - 0x{:X} (0x{:X})", map.data(), map.end().base(), map.size());
+        if (guest.valid())
+            throw exception("Mapping KSharedMemory multiple times on guest is not supported: Requested Mapping: 0x{:X} - 0x{:X} (0x{:X}), Current Mapping: 0x{:X} - 0x{:X} (0x{:X})", map.data(), map.end().base(), map.size(), guest.data(), guest.end().base(), guest.size());
 
-        guest.ptr = static_cast<u8 *>(mmap(ptr, size, permission.Get(), MAP_SHARED | (ptr ? MAP_FIXED : 0), fd, 0));
-        if (guest.ptr == MAP_FAILED)
+        auto guest_ptr = static_cast<u8 *>(mmap(map.data(), map.size(), permission.Get(), MAP_SHARED | (map.data() ? MAP_FIXED : 0), fd, 0));
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp:33]: guestPtr

**Review by bylaws — CHANGES_REQUESTED** [review]
this code is pretty cute

File: app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp [hunk 8]
```diff
@@ -10,76 +10,76 @@
...
+        if (!state.process->memory.base.contains(map))
+            throw exception("KPrivateMemory allocation isn't inside guest address space: 0x{:X} - 0x{:X}", map.data(), map.end().base());
+        if (!util::IsPageAligned(map.data()) || !util::IsPageAligned(map.size()))
+            throw exception("KSharedMemory mapping isn't page-aligned: 0x{:X} - 0x{:X} (0x{:X})", map.data(), map.end().base(), map.size());
+        if (guest.valid())
+            throw exception("Mapping KSharedMemory multiple times on guest is not supported: Requested Mapping: 0x{:X} - 0x{:X} (0x{:X}), Current Mapping: 0x{:X} - 0x{:X} (0x{:X})", map.data(), map.end().base(), map.size(), guest.data(), guest.end().base(), guest.size());
 
-        guest.ptr = static_cast<u8 *>(mmap(ptr, size, permission.Get(), MAP_SHARED | (ptr ? MAP_FIXED : 0), fd, 0));
-        if (guest.ptr == MAP_FAILED)
+        auto guestPtr = static_cast<u8 *>(mmap(map.data(), map.size(), permission.Get(), MAP_SHARED | (map.data() ? MAP_FIXED : 0), fd, 0));
```
Comment (bylaws) [inline app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp:33]: use {}

**Review by PixelyIon — CHANGES_REQUESTED** [review]
There is a very large amount of nits to fix here, it'll take you a while. Good luck!

**Review by PixelyIon — APPROVED** [review]
Great job, LGTM! I'll merge after you rebase onto `ftx1`.

**Review by bylaws — APPROVED** [review]

**Pull request merged**